pkg/awsiamauth/reconciler/reconciler.go

```diff
@@ -0,0 +1,161 @@
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("generating aws-iam-authenticator ca secret: %v", err)
+	}
+
+	return controller.Result{}, serverside.ReconcileYaml(ctx, r.client, yaml)
+}
+
+// Reconcile takes the AWS IAM Authenticator installation to the desired state defined in AWSIAMConfig.
+// It uses a controller.Result to indicate when requeues are needed.
+// Intended to be used in a kubernetes controller.
+func (r *Reconciler) Reconcile(ctx context.Context, logger logr.Logger, cluster *anywherev1.Cluster) (controller.Result, error) {
+	clusterSpec, err := anywhereCluster.BuildSpec(ctx, clientutil.NewKubeClient(r.client), cluster)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	result, err := clusters.CheckControlPlaneReady(ctx, r.client, logger, cluster)
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("checking controlplane ready: %v", err)
+	}
+
+	if result.Return() {
+		return result, nil
+	}
+
+	if err := r.ensureCASecretOwnerRef(ctx, logger, cluster); err != nil {
+		return controller.Result{}, err
+	}
+
+	rClient, err := r.remoteClientRegistry.GetClient(ctx, controller.CapiClusterObjectKey(cluster))
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	var clusterID uuid.UUID
+	cm := &corev1.ConfigMap{}
+	err = rClient.Get(ctx, types.NamespacedName{Name: awsiamauth.AwsIamAuthConfigMapName, Namespace: constants.KubeSystemNamespace}, cm)
+	if apierrors.IsNotFound(err) {
+		// If configmap is not found, this is a first time install of aws-iam-authenticator on the cluster.
+		// We use a newly generated UUID.
+		clusterID = r.generateUUID()
+	} else if err != nil {
+		return controller.Result{}, fmt.Errorf("fetching configmap %s: %v", awsiamauth.AwsIamAuthConfigMapName, err)
+	}
+
+	logger.Info("Applying aws-iam-authenticator manifest")
+	return r.applyIAMAuthManifest(ctx, rClient, clusterSpec, clusterID)
+}
+
+func (r *Reconciler) applyIAMAuthManifest(ctx context.Context, client client.Client, clusterSpec *anywhereCluster.Spec, clusterID uuid.UUID) (controller.Result, error) {
+	yaml, err := r.templateBuilder.GenerateManifest(clusterSpec, clusterID)
+	if err != nil {
+		return controller.Result{}, fmt.Errorf("generating aws-iam-authenticator manifest: %v", err)
+	}
+
+	return controller.Result{}, serverside.ReconcileYaml(ctx, client, yaml)
+}
+
+// ensureCASecretOwnerRef ensures that the CAPI Cluster object is set as an ownerReference.
+// The ownerReference ensures the CA Secret is deleted when the cluster is deleted.
+func (r *Reconciler) ensureCASecretOwnerRef(ctx context.Context, logger logr.Logger, cluster *anywherev1.Cluster) error {
+	secret := &corev1.Secret{}
+	if err := r.client.Get(ctx, types.NamespacedName{Name: awsiamauth.CASecretName(cluster.Name), Namespace: constants.EksaSystemNamespace}, secret); err != nil {
+		return fmt.Errorf("fetching secret %s: %v", awsiamauth.CASecretName(cluster.Name), err)
+	}
+
+	if len(secret.GetOwnerReferences()) != 0 {
```

Comment:
this is not safe, what if the secret already has a different owner ref?

Reply:
Fair point. Will do a previous count comparison.